### ISSUE TITLE
TestBase in owlapi-contract directly uses Impl classes from owlapi-impl  

### ISSUE THREAD

**Issue by sesuncedu** [owlcs/owlapi#604]
I had the general impression that the original reason for having tests in a separate owlapi-contract subproject was to test conformance to the public API, without tying in to a specific implementation.  This is a good idea, and helps with modularity. 

TestBase behaves oddly in that it uses OWLManager to create a prototype ontology manager, but then directly creates instances of OWLOntologyManagerImpl and copies the information into them. 

Comments in the log suggest that this was done in revision 7f75809543915977e4fae782d371185388a2bfd6 to avoid injection costs. 
The injection tuning discussed in issue #580 may make this unnecessary

TestBase also creates a custom builder with a direct reference to OWLOntologyImpl 

This field is used in only a few places:
* org.semanticweb.owlapi.api.test.syntax.rdf.DisjointsTestCase
* org.semanticweb.owlapi.api.test.syntax.rdf.RendererAndParserTestCase
* org.semanticweb.owlapi.api.test.syntax.rdfxml.RDFParserTestCase

In each of the builder used to construct a new ontology factory, replacing all others (and creating another dependency on OWLOntologyFactoryImpl).  Is the default unsuitable?

```java
        m.getOntologyFactories().set(new OWLOntologyFactoryImpl(builder));
```

**Comment by sesuncedu** [comment]
There are also a few references to the prototype three test cases ,where it is used as a hack to modify  the configurator when using convenience methods that don't give access to the configuration.

A hack to avoid this kludge is to make the ontology creation method overridable, and have the test cases that were using this kludge override the method.

**Comment by sesuncedu** [comment]
Side note:  tests didn't seem noticeably slower, so tuning mentioned in #580 seems to work.

**Comment by ignazio1977** [comment]
All of those kludges were originally written to allow tests to run faster. If dropping them no longer affects the test speed, then it's a good thing to remove them.

**Comment by ignazio1977** [comment]
I've tried removing the kludge but the test time doubled on my machine. I think we need to review this when the injectors are seriously rejiggled.

**Comment by sesuncedu** [comment]
I wonder if this is related to the test runner set up?
A microbenchmark might be useful (also, I need to figure out how to set up the gradle plugin for jmh  benchmarks)

**Comment by ignazio1977** [comment]
I think it would be too much effort for little gain at the moment, given that there's half a plan of switching to something like Dagger.

The doubling is still in line with my previous profiling results - I saw a 60% improvement from my previous set of tweaks in the injector modules, and the previous slowdown from injection was a 4-5x increase - from a bit more than one minute to about five. The doubling is pretty close to what I'd extrapolate, so I don't think the test setup is sensibly different from the cost required for creating a large number of managers.

**Comment by sesuncedu** [comment]
It's worthwhile precisely because of the possible daggering.

**Comment by sesuncedu** [comment]
As I predicted to myself, the benchmark code was trivial;  getting the
gradle plugin to behave was much harder.

With no warmup iterations, and forking a fresh virtual machine for each
iteration, with 32 datapoints a call to
OWLManager.createOWLOntologyMananger,  a completely cold call to
createOWLOntologyMananger has the following characteristics:

  0.418 ±(99.9%) 0.140 s/op [Average]
  (min, avg, max) = (0.252, 0.418, 1.029), stdev = 0.217
  CI (99.9%): [0.278, 0.557] (assumes normal distribution)

With a single warmup iteration, the times change to:

Result "createOWLOntologyManager":
  0.015 ±(99.9%) 0.001 s/op [Average]
  (min, avg, max) = (0.014, 0.015, 0.021), stdev = 0.001
  CI (99.9%): [0.015, 0.016] (assumes normal distribution)

With four warmup iterations, this becomes

Result "createOWLOntologyManager":
  0.004 ±(99.9%) 0.001 s/op [Average]
  (min, avg, max) = (0.004, 0.004, 0.005), stdev = 0.001
  CI (99.9%): [0.004, 0.004] (assumes normal distribution)

This seems to suggest  that the time taken to create multiple owl ontology
managers in a new VM  is dominated by the time taken to create the first
manager. It is possible that there are other costs that are not included in
the simple call to cOOM.

**Comment by sesuncedu** [comment]
Adding a second benchmark that creates the OOM, and counts the number of ontology,parser, and storer factories does not indicate major differences:
with 0 warmups
```
Benchmark                                                Mode  Cnt  Score   Error  Units
OntologyManagerCreateBenchmark.countFactories            avgt   10  0.527 ± 0.299   s/op
OntologyManagerCreateBenchmark.createOWLOntologyManager  avgt   10  0.503 ± 0.438   s/op
```
and with 4 warmups
```
OntologyManagerCreateBenchmark.countFactories            avgt   10  0.004 ±  0.001   s/op
OntologyManagerCreateBenchmark.createOWLOntologyManager  avgt   10  0.004 ±  0.001   s/op
```

**Comment by ignazio1977** [comment]
Fixed in 0e3c6b921e401dbee1a29efd3c3302a1fb8628a2 - just allow the injector to be cached by the calling application. This will also allow third parties to write fast tests. #580 @ykazakov

**Comment by ignazio1977** [comment]
Fixed in 7e695439ff2dfc5f68f6ddd9a9bbfcc901045be1 for version 4
